File: 0085-maximal-rectangle/0085-maximal-rectangle.java
class Solution {
    public int maximalRectangle(char[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            return 0;
        }

        int maxArea = 0;
        int dp[] = new int [matrix[0].length];

        for(int i = 0; i< matrix.length; i++){
            for(int j =0; j< matrix[0].length; j++){
                dp[j] = matrix[i][j] =='1' ? dp[j]+1 : 0;
            }
            maxArea = Math.max(maxArea , maxHistogram(dp));
        }
       
        return maxArea;
    }

    public int maxHistogram(int dp[]){
        Stack<int[]> stack = new Stack<>();
        int n = dp.length;
        int maxArea = 0;

        for(int i = 0;i < n ; i++ ){
            int h =dp[i];
            int start = i;
            while(!stack.isEmpty() && stack.peek()[1] > h){
                int popped[] = stack.pop();
                int index = popped[0];
                int height = popped[1];
                maxArea = Math.max(maxArea , height * (i - index));
                start = index;
            }
            stack.push(new int[]{start, h});
        }

        while(!stack.isEmpty()){
            int popped[] = stack.pop();
            int index = popped[0];
            int height = popped[1];
            maxArea = Math.max(maxArea , height * (dp.length - index));
        }
        return maxArea;
    }
}

File: 0085-maximal-rectangle/0085-maximal-rectangle.py
class Solution:
    def maxHistogram(self,heights: List[int])->int:
        stack = [] # pair of index, height
        maxArea = 0

        for i,h in enumerate(heights):
            start = i
            while stack and stack[-1][1] >h:
                index,height = stack.pop()
                maxArea = max(maxArea , height * (i - index))
                start = index
            stack.append((start,h))

        for index, height in stack:
            maxArea = max(maxArea, height *(len(heights)-index))
        
        return maxArea

    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        if not matrix or not matrix[0]:
            return 0
        m = len(matrix)
        n = len(matrix[0])
        maxArea = 0

        heights = [0]*n

        for i in range(m):
            for j in range(n):
                if matrix[i][j]=='1':
                    heights[j] += 1
                else:
                    heights[j] = 0
            maxArea = max(maxArea , self.maxHistogram(heights))
        
        return maxArea

    
